src/Compilers/CSharp/Portable/Binder/Imports.cs

```diff
@@ -252,16 +272,124 @@ public static Imports FromGlobalUsings(CSharpCompilation compilation)
                 }
             }
 
-            uniqueUsings.Free();
-
             if (diagnostics.IsEmptyWithoutResolution)
             {
                 diagnostics = null;
             }
 
+            var previousSubmissionImports = compilation.PreviousSubmission?.GlobalImports;
+            if (previousSubmissionImports != null)
+            {
+                // Currently, only usings are supported.
+                Debug.Assert(previousSubmissionImports.UsingAliases.IsEmpty);
+                Debug.Assert(previousSubmissionImports.ExternAliases.IsEmpty);
+
+                var expandedImports = ExpandPreviousSubmissionImports(previousSubmissionImports, compilation);
+
+                foreach (var previousUsing in expandedImports.Usings)
+                {
+                    if (uniqueUsings.Add(previousUsing.NamespaceOrType))
+                    {
+                        boundUsings.Add(previousUsing);
+                    }
+                }
+            }
+
+            uniqueUsings.Free();
+
+            if (boundUsings.Count == 0)
+            {
+                boundUsings.Free();
+                return Empty;
+            }
```

Comment:
Here it makes almost no difference.  Downstream, things like ```Concat``` have early outs for ```Empty```.  I almost applied the optimization globally (e.g. by using a factory method rather than a constructor) but didn't want to clutter the PR.